cs/src/Connections/TunnelClient.cs

```diff
@@ -254,6 +258,43 @@ private void ConfigurePortForwardingService()
         {
             pfs.MessageFactory = this;
             pfs.ForwardedPortConnecting += OnForwardedPortConnecting;
+            pfs.RemoteForwardedPorts.PortAdded += (_, e) => OnForwardedPortAdded(pfs, e);
+            pfs.RemoteForwardedPorts.PortUpdated += (_, e) => OnForwardedPortAdded(pfs, e);
+        }
+    }
+
+    private void OnForwardedPortAdded(PortForwardingService pfs, ForwardedPortEventArgs e)
+    {
+        var port = e.Port.RemotePort;
+        if (!port.HasValue)
+        {
+            return;
+        }
+
+        var disconnectedStreamsCount = 0;
+        lock (this.disconnectedStreams)
+        {
+            // If there are disconnected streams for the port, re-connect them now.
+            if (this.disconnectedStreams.TryGetValue(port.Value, out var streams))
```

Comment:
Are we supposed to remove the port from `disconnectedStreams` if reconnecting the streams is successful?

Reply:
That is done on line 339.

This method restores channels for the forwarded port, then the `OnForwardedPortConnectiong()` method below reconnects the `SecureStream` for each channel.